File: c_sharp/lesson19/lesson19.cs
/* 
 * NeHe Lesson 19 - Particle Engine using Triangle Strips
 * http://nehe.gamedev.net/data/lessons/lesson.asp?lesson=19
 * 
 * Ported to C# by Brian Holley, February, 2004
 * http://tachyon.unl.edu
 * 
 * Uses the CsGL library from http://csgl.sourceforge.net
 * Some CsGL setup and image code from NeHe Lesson 06 C# port by Sabine Felsinger
 */

using System;
using System.Drawing;
using System.Windows.Forms;
using CsGL.OpenGL;

namespace Lesson19
{
	public class MainForm : System.Windows.Forms.Form	
	{
		public Lesson19.LessonView view;

		public MainForm()
		{
			this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
			this.ClientSize = new System.Drawing.Size(640, 480);
			this.Name = "MainForm";
			this.Text = "NeHe Lesson 19 - C#";
			this.view = new Lesson19.LessonView();
			this.view.Parent = this;
			this.view.Dock = DockStyle.Fill;
			this.Show();
		}

		static void Main() 
		{
			MainForm form = new MainForm();

			while ((!form.view.finished) && (!form.IsDisposed))
			{
				form.view.glDraw();
				form.Refresh();
				Application.DoEvents();
			}

			form.Dispose();
		}
	}
	
	public class LessonView : OpenGLControl
	{
		public static int MaxParticles = 500;	// Number of particles to create
		public bool	rainbow = true;				// Rainbow Mode?
		public bool	sp = false;					// Spacebar Pressed?
		public bool	rp = false;					// Enter Key Pressed?

		public float slowdown = 2.0f;			// Slow Down Particles
		public float xspeed = 0.0f;				// Base X Speed (To Allow Keyboard Direction Of Tail)
		public float yspeed = 0.0f;				// Base Y Speed (To Allow Keyboard Direction Of Tail)
		public float zoom = -40.0f;				// Used To Zoom Out

		public uint col = 0;					// Current Color Selection
		public uint delay = 0;					// Rainbow Effect Delay
		public Random rand = null;				// Random number generator
		
		public uint[] texture = new uint[3];	// Texture array

		public class Particle					// Create A Structure For Particle
		{
			public bool	active;					// Active (Yes/No)
			public float life;					// Particle Life
			public float fade;					// Fade Speed
			public float r, g, b;				// Color
			public float x, y, z;				// Position
			public float xi, yi, zi;			// Direction
			public float xg, yg, zg;			// Gravity
		}
		
		Particle[] particle = null;				// Particle Array (Room For Particle Info)

		public static float[][] colors = new float[12][] {	// Rainbow Of Colors
			new float[3] {1.0f,0.5f,0.5f}, new float[3] {1.0f,0.75f,0.5f}, 
			new float[3] {1.0f,1.0f,0.5f}, new float[3] {0.75f,1.0f,0.5f},
			new float[3] {0.5f,1.0f,0.5f}, new float[3] {0.5f,1.0f,0.75f}, 
			new float[3] {0.5f,1.0f,1.0f}, new float[3] {0.5f,0.75f,1.0f},
			new float[3] {0.5f,0.5f,1.0f}, new float[3] {0.75f,0.5f,1.0f}, 
			new float[3] {1.0f,0.5f,1.0f}, new float[3] {1.0f,0.5f,0.75f} };

		public bool finished;
        
		public LessonView() : base()
		{
			this.KeyDown += new KeyEventHandler(LessonView_KeyDown);
			this.KeyUp += new KeyEventHandler(LessonView_KeyUp);
			this.finished = false;

			this.particle = new Particle[LessonView.MaxParticles];
			for (int i=0; i < LessonView.MaxParticles; i++)
				this.particle[i] = new Particle();
		}

		protected override void InitGLContext() 
		{
			LoadTextures();

			GL.glEnable(GL.GL_TEXTURE_2D);									// Enable Texture Mapping
			GL.glShadeModel(GL.GL_SMOOTH);									// Enable Smooth Shading
			GL.glClearColor(0.0f, 0.0f, 0.0f, 0.5f);						// Black Background
			GL.glClearDepth(1.0f);											// Depth Buffer Setup
			GL.glDisable(GL.GL_DEPTH_TEST);									// Enables Depth Testing
			GL.glEnable(GL.GL_BLEND);										// Enable Blending
			GL.glBlendFunc(GL.GL_SRC_ALPHA, GL.GL_ONE);						// Type Of Blending To Perform
			GL.glHint(GL.GL_PERSPECTIVE_CORRECTION_HINT, GL.GL_NICEST);		// Really Nice Perspective Calculations

			GL.glHint(GL.GL_POINT_SMOOTH_HINT, GL.GL_NICEST);				// Really Nice Point Smoothing

			GL.glBindTexture(GL.GL_TEXTURE_2D, this.texture[0]);			// Select Our Texture

			this.rand = new Random();
			for (int loop=0; loop < LessonView.MaxParticles; loop++)		// Initials All The Textures
			{
				this.particle[loop].active = true;								// Make All The Particles Active
				this.particle[loop].life = 1.0f;								// Give All The this.particles Full Life
				this.particle[loop].fade = (float)(this.rand.Next(100))/1000.0f+0.003f;	// Random Fade Speed
				this.particle[loop].r = LessonView.colors[loop*(12/LessonView.MaxParticles)][0];	// Select Red Rainbow Color
				this.particle[loop].g = LessonView.colors[loop*(12/LessonView.MaxParticles)][1];	// Select Red Rainbow Color
				this.particle[loop].b = LessonView.colors[loop*(12/LessonView.MaxParticles)][2];	// Select Red Rainbow Color
				this.particle[loop].xi = (float)((this.rand.Next(50))-26.0f)*10.0f;	// Random Speed On X Axis
				this.particle[loop].yi = (float)((this.rand.Next(50))-25.0f)*10.0f;	// Random Speed On Y Axis
				this.particle[loop].zi = (float)((this.rand.Next(50))-25.0f)*10.0f;	// Random Speed On Z Axis
				this.particle[loop].xg = 0.0f;									// Set Horizontal Pull To Zero
				this.particle[loop].yg = -0.8f;									// Set Vertical Pull Downward
				this.particle[loop].zg = 0.0f;									// Set Pull On Z Axis To Zero
			}
		}

		protected bool LoadTextures()
		{
			Bitmap image = null;
			string file = @"Data\Particle.bmp";
			try
			{
				// If the file doesn't exist or can't be found, an ArgumentException is thrown instead of
				// just returning null
				image = new Bitmap(file);
			} 
			catch (System.ArgumentException)
			{
				MessageBox.Show("Could not load " + file + ".  Please make sure that Data is a subfolder from where the application is running.", "Error", MessageBoxButtons.OK);
				this.finished = true;
			}
			if (image != null)
			{
				image.RotateFlip(RotateFlipType.RotateNoneFlipY);
				System.Drawing.Imaging.BitmapData bitmapdata;
				Rectangle rect = new Rectangle(0, 0, image.Width, image.Height);

				bitmapdata = image.LockBits(rect, System.Drawing.Imaging.ImageLockMode.ReadOnly, System.Drawing.Imaging.PixelFormat.Format24bppRgb);

				GL.glGenTextures(3, this.texture);
			
				// Create Nearest Filtered Texture
				GL.glBindTexture(GL.GL_TEXTURE_2D, this.texture[0]);
				GL.glTexParameteri(GL.GL_TEXTURE_2D, GL.GL_TEXTURE_MAG_FILTER, GL.GL_NEAREST);
				GL.glTexParameteri(GL.GL_TEXTURE_2D, GL.GL_TEXTURE_MIN_FILTER, GL.GL_NEAREST);
				GL.glTexImage2D(GL.GL_TEXTURE_2D, 0, (int)GL.GL_RGB, image.Width, image.Height, 0, GL.GL_BGR_EXT, GL.GL_UNSIGNED_BYTE, bitmapdata.Scan0);

				// Create Linear Filtered Texture
				GL.glBindTexture(GL.GL_TEXTURE_2D, this.texture[1]);
				GL.glTexParameteri(GL.GL_TEXTURE_2D, GL.GL_TEXTURE_MAG_FILTER, GL.GL_LINEAR);
				GL.glTexParameteri(GL.GL_TEXTURE_2D, GL.GL_TEXTURE_MIN_FILTER, GL.GL_LINEAR);
				GL.glTexImage2D(GL.GL_TEXTURE_2D, 0, (int)GL.GL_RGB, image.Width, image.Height, 0, GL.GL_BGR_EXT, GL.GL_UNSIGNED_BYTE, bitmapdata.Scan0);

				// Create MipMapped Texture
				GL.glBindTexture(GL.GL_TEXTURE_2D, this.texture[2]);
				GL.glTexParameteri(GL.GL_TEXTURE_2D, GL.GL_TEXTURE_MAG_FILTER, GL.GL_LINEAR);
				GL.glTexParameteri(GL.GL_TEXTURE_2D, GL.GL_TEXTURE_MIN_FILTER, GL.GL_LINEAR_MIPMAP_NEAREST);
				GL.gluBuild2DMipmaps(GL.GL_TEXTURE_2D, (int)GL.GL_RGB, image.Width, image.Height, GL.GL_BGR_EXT, GL.GL_UNSIGNED_BYTE, bitmapdata.Scan0);

				image.UnlockBits(bitmapdata);
				image.Dispose();
				return true;
			}
			return false;
		}

		public override void glDraw()
		{
			GL.glClear(GL.GL_COLOR_BUFFER_BIT | GL.GL_DEPTH_BUFFER_BIT);
			GL.glLoadIdentity();
			
			for (int loop=0; loop < LessonView.MaxParticles; loop++)	// Loop Through All The Particles
			{
				if (this.particle[loop].active)							// If The Particle Is Active
				{
					float x = this.particle[loop].x;					// Grab Our Particle X Position
					float y = this.particle[loop].y;					// Grab Our Particle Y Position
					float z = this.particle[loop].z+zoom;				// Particle Z Pos + Zoom

					// Draw The Particle Using Our RGB Values, Fade The Particle Based On It's Life
					GL.glColor4f(this.particle[loop].r, this.particle[loop].g, this.particle[loop].b, this.particle[loop].life);

					GL.glBegin(GL.GL_TRIANGLE_STRIP);						// Build Quad From A TrianGL.gle Strip
						GL.glTexCoord2d(1, 1); GL.glVertex3f(x+0.5f, y+0.5f, z); // Top Right
						GL.glTexCoord2d(0, 1); GL.glVertex3f(x-0.5f, y+0.5f, z); // Top Left
						GL.glTexCoord2d(1, 0); GL.glVertex3f(x+0.5f, y-0.5f, z); // Bottom Right
						GL.glTexCoord2d(0, 0); GL.glVertex3f(x-0.5f, y-0.5f, z); // Bottom Left
					GL.glEnd();										// Done Building TrianGL.gle Strip

					this.particle[loop].x += this.particle[loop].xi/(slowdown*1000);// Move On The X Axis By X Speed
					this.particle[loop].y += this.particle[loop].yi/(slowdown*1000);// Move On The Y Axis By Y Speed
					this.particle[loop].z += this.particle[loop].zi/(slowdown*1000);// Move On The Z Axis By Z Speed

					this.particle[loop].xi += this.particle[loop].xg;			// Take Pull On X Axis Into Account
					this.particle[loop].yi += this.particle[loop].yg;			// Take Pull On Y Axis Into Account
					this.particle[loop].zi += this.particle[loop].zg;			// Take Pull On Z Axis Into Account
					this.particle[loop].life -= this.particle[loop].fade;		// Reduce Particles Life By 'Fade'

					if (this.particle[loop].life < 0.0f)					// If Particle Is Burned Out
					{
						this.particle[loop].life = 1.0f;					// Give It New Life
						this.particle[loop].fade = (float)(this.rand.Next(100))/1000.0f+0.003f;	// Random Fade Value
						this.particle[loop].x = 0.0f;						// Center On X Axis
						this.particle[loop].y = 0.0f;						// Center On Y Axis
						this.particle[loop].z = 0.0f;						// Center On Z Axis
						this.particle[loop].xi = this.xspeed+(float)(this.rand.Next(60)-32.0f);	// X Axis Speed And Direction
						this.particle[loop].yi = this.yspeed+(float)(this.rand.Next(60)-30.0f);	// Y Axis Speed And Direction
						this.particle[loop].zi = (float)(this.rand.Next(60)-30.0f);	// Z Axis Speed And Direction
						this.particle[loop].r = LessonView.colors[this.col][0];			// Select Red From Color Table
						this.particle[loop].g = LessonView.colors[this.col][1];			// Select Green From Color Table
						this.particle[loop].b = LessonView.colors[this.col][2];			// Select Blue From Color Table
					}
				}
			}

			if (this.rainbow && (this.delay > 25))
			{
				this.delay = 0;						// Reset The Rainbow Color Cycling Delay
				this.col = (this.col + 1) % 12;		// Change The Particle Color
			}

			delay++;							// Increase Rainbow Mode Color Cycling Delay Counter
		}

		protected override void OnSizeChanged(EventArgs e)
		{
			base.OnSizeChanged(e);
			Size s = Size;

			if (s.Height == 0)
				s.Height = 1;

			GL.glViewport(0, 0, s.Width, s.Height);

			GL.glMatrixMode(GL.GL_PROJECTION);
			GL.glLoadIdentity();
			GL.gluPerspective(45.0f, (double)s.Width /(double) s.Height, 0.1f, 100.0f);	
			GL.glMatrixMode(GL.GL_MODELVIEW);
			GL.glLoadIdentity();
		}

		protected void LessonView_KeyDown(object Sender, KeyEventArgs e)
		{
			if (e.KeyCode == Keys.Escape)				// Finish the application if the escape key was pressed
				this.finished = true;
			else if (e.KeyCode == Keys.Add && (this.slowdown > 1.0f))
				this.slowdown -= 0.01f;		// Speed Up Particles
			else if (e.KeyCode == Keys.Subtract && (this.slowdown < 4.0f))
				this.slowdown+=0.01f;	// Slow Down Particles
			else if (e.KeyCode == Keys.PageUp)
				this.zoom += 0.1f;		// Zoom In
			else if (e.KeyCode == Keys.PageDown)
				this.zoom -= 0.1f;		// Zoom Out
			else if (e.KeyCode == Keys.Return && !this.rp)	// Return Key Pressed
			{
				this.rp = true;								// Set Flag Telling Us It's Pressed
				this.rainbow = !this.rainbow;				// Toggle Rainbow Mode On / Off
			}	
			else if (e.KeyCode == Keys.Space && !this.sp)	// Space Or Rainbow Mode
			{
				this.rainbow = false;			// If Spacebar Is Pressed Disable Rainbow Mode
				this.sp = true;						// Set Flag Telling Us Space Is Pressed
				this.delay = 0;						// Reset The Rainbow Color Cycling Delay
				this.col = (this.col + 1) % 12;		// Change The Particle Color
			}
			else
			{
				for (int loop=0; loop < LessonView.MaxParticles; loop++)
				{
					// If Number Pad 8 And Y Gravity Is Less Than 1.5 Increase Pull Upwards
					if (e.KeyCode == Keys.NumPad8 && (this.particle[loop].yg < 1.5f))
						this.particle[loop].yg += 0.01f;
						// If Number Pad 2 And Y Gravity Is Greater Than -1.5 Increase Pull Downwards
					else if (e.KeyCode == Keys.NumPad2 && (this.particle[loop].yg > -1.5f))
						this.particle[loop].yg -= 0.01f;
						// If Number Pad 6 And X Gravity Is Less Than 1.5 Increase Pull Right
					else if (e.KeyCode == Keys.NumPad6 && (this.particle[loop].xg < 1.5f))
						this.particle[loop].xg += 0.01f;
						// If Number Pad 4 And X Gravity Is Greater Than -1.5 Increase Pull Left
					else if (e.KeyCode == Keys.NumPad4 && (this.particle[loop].xg > -1.5f))
						this.particle[loop].xg -= 0.01f;
					else if (e.KeyCode == Keys.Tab)										// Tab Key Causes A Burst
					{
						this.particle[loop].x = 0.0f;								// Center On X Axis
						this.particle[loop].y = 0.0f;								// Center On Y Axis
						this.particle[loop].z = 0.0f;								// Center On Z Axis
						this.particle[loop].xi = (float)(this.rand.Next(50)-26.0f)*10.0f;	// Random Speed On X Axis
						this.particle[loop].yi = (float)(this.rand.Next(50)-25.0f)*10.0f;	// Random Speed On Y Axis
						this.particle[loop].zi = (float)(this.rand.Next(50)-25.0f)*10.0f;	// Random Speed On Z Axis
					}
				}
			}
		}

		private void LessonView_KeyUp(object sender, KeyEventArgs e)
		{
			if (e.KeyCode == Keys.Space)			// Release the space key lock
				this.sp = false;
			else if (e.KeyCode == Keys.Return)		// Release the return key lock
				this.rp = false;
		}

		protected override bool ProcessDialogKey(Keys keyData)
		{
			if (keyData == Keys.Up && this.yspeed < 200)						// Change rotation about the x axis
				this.yspeed += 1.0f;
			else if (keyData == Keys.Down && this.yspeed > -200)
				this.yspeed -= 1.0f;
			else if (keyData == Keys.Right && this.xspeed < 200)				// Change rotation about the y axis
				this.xspeed += 1.0f;
			else if (keyData == Keys.Left && this.xspeed > -200)
				this.xspeed -= 1.0f;
		
			return base.ProcessDialogKey (keyData);
		}
	}
}

File: irix/lesson19/Lesson19.c
///////////////////////////////////////////////////////////////////////////////
//    Lesson 19 code by Jeff Molofee 
//    Modified to run on Irix using GLUT by Dimitrios Christopoulos      
//    Date: 28/03/00
///////////////////////////////////////////////////////////////////////////////
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <GL/gl.h>
#include <GL/glu.h>
#include <GL/glut.h>
#include "texture.h"


#define	MAX_PARTICLES	1000		// Number Of Particles To Create

int	keys[256];					// Array Used For The Keyboard Routine
int	active=1;				// Window Active Flag Set To 1 By Default
int	rainbow=0;				// Rainbow Mode?
float   df=0.2, df1=0.01, df2=-0.02;

float	slowdown=2.0f;				// Slow Down Particles
float	xspeed;						// Base X Speed (To Allow Keyboard Direction Of Tail)
float	yspeed;						// Base Y Speed (To Allow Keyboard Direction Of Tail)
float	zoom=-40.0f;				// Used To Zoom Out
float   spin=0, spin1=0, spin2=0;

GLuint	loop;						// Misc Loop Variable
GLuint	col;						// Current Color Selection
GLuint	delay=0;						// Rainbow Effect Delay
GLuint	image;						// Storage For Particle Image Display List
GLuint	texture[1];					// Storage For Our Particle Texture

typedef struct						// Create A Structure For Particle
{
	int	active;					// Active (Yes/No)
	float	life;					// Particle Life
	float	fade;					// Fade Speed
	float	r;						// Red Value
	float	g;						// Green Value
	float	b;						// Blue Value
	float	x;						// X Position
	float	y;						// Y Position
	float	z;						// Z Position
	float	xi;						// X Direction
	float	yi;						// Y Direction
	float	zi;						// Z Direction
	float	xg;						// X Gravity
	float	yg;						// Y Gravity
	float	zg;						// Z Gravity
}
particles;							// Particles Structure

particles particle[MAX_PARTICLES];	// Particle Array (Room For Particle Info)

static GLfloat colors[12][3]=		// Rainbow Of Colors
{
	{1.0f,0.5f,0.5f},{1.0f,0.75f,0.5f},{1.0f,1.0f,0.5f},{0.75f,1.0f,0.5f},
	{0.5f,1.0f,0.5f},{0.5f,1.0f,0.75f},{0.5f,1.0f,1.0f},{0.5f,0.75f,1.0f},
	{0.5f,0.5f,1.0f},{0.75f,0.5f,1.0f},{1.0f,0.5f,1.0f},{1.0f,0.5f,0.75f}
};


//Loads texture , the textures have to be in sgi(rgb) format
//I used the texture loading code code from the advanced siggraff 99 notes
// this code is in the texture.c .h files,  and is fairly easy to use.
int LoadGLTextures()								// Load Bitmap And Convert To Textures
{
        int Status=0;								// Status Indicator
        unsigned *TextureImage=NULL;				// Create Storage Space For The Textures
        int texwid, texht; /* dimensions of texture that was read */
	int texcomps; /* number of components in external format */

        if (TextureImage=read_texture("Data/Particle.rgb", &texwid, &texht, &texcomps)) // Load Particle Texture
        {
			Status=1;								// Set The Status To 1
			glGenTextures(1, &texture[0]);				// Create One Texture

			glBindTexture(GL_TEXTURE_2D, texture[0]);
			glTexParameteri(GL_TEXTURE_2D,GL_TEXTURE_MAG_FILTER,GL_LINEAR);
			glTexParameteri(GL_TEXTURE_2D,GL_TEXTURE_MIN_FILTER,GL_LINEAR);
			glTexImage2D(GL_TEXTURE_2D, 0, texcomps,texwid,texht, 0, GL_RGBA, GL_UNSIGNED_BYTE, TextureImage);
        }

        if (TextureImage)							// If Texture Exists
		{
		 	free(TextureImage);			// Free The Texture Image Memory
		}
               
        return Status;									// Return The Status
}

GLvoid MakeParticle(GLvoid)								// Build Our Particle Display List
{
	image=glGenLists(1);								// Generate Our Display List
	glNewList(image,GL_COMPILE);						// Build A Compiled Display List
		glBegin(GL_QUADS);								// Build A Quad
			glTexCoord2f(1.0f,1.0f); glVertex3f( 0.5f, 0.5f, 0.0f);	// Top Right
			glTexCoord2f(0.0f,1.0f); glVertex3f(-0.5f, 0.5f, 0.0f);	// Top Left
			glTexCoord2f(0.0f,0.0f); glVertex3f(-0.5f,-0.5f, 0.0f);	// Bottom Left
			glTexCoord2f(1.0f,0.0f); glVertex3f( 0.5f,-0.5f, 0.0f);	// Bottom Right
		glEnd();										// Done Building Our Quad
	glEndList();										// Done Building Our Display List
}

void reshape(int width, int height)		// Resize And Initialize The GL Window
{
	if (height==0)										// Prevent A Divide By Zero By
	{
		height=1;										// Making Height Equal One
	}

	glViewport(0,0,width,height);						// Reset The Current Viewport

	glMatrixMode(GL_PROJECTION);						// Select The Projection Matrix
	glLoadIdentity();									// Reset The Projection Matrix

	// Calculate The Aspect Ratio Of The Window
	gluPerspective(45.0f,(GLfloat)width/(GLfloat)height,0.1f,200.0f);

	glMatrixMode(GL_MODELVIEW);							// Select The Modelview Matrix
	glLoadIdentity();									// Reset The Modelview Matrix
}

int InitGL(GLvoid)										// All Setup For OpenGL Goes Here
{
	if (!LoadGLTextures())								// Jump To Texture Loading Routine
	{
		return 0;									// If Texture Didn't Load Return 0
	}

	MakeParticle();										// Build Our Particle Display List

	glShadeModel(GL_SMOOTH);							// Enable Smooth Shading
	glClearColor(0.0f,0.0f,0.0f,0.0f);					// Black Background
	glClearDepth(1.0f);									// Depth Buffer Setup
	glDisable(GL_DEPTH_TEST);							// Disable Depth Testing
	glEnable(GL_BLEND);									// Enable Blending
	glBlendFunc(GL_SRC_ALPHA,GL_ONE);					// Type Of Blending To Perform
	glHint(GL_PERSPECTIVE_CORRECTION_HINT,GL_NICEST);	// Really Nice Perspective Calculations
	glHint(GL_POINT_SMOOTH_HINT,GL_NICEST);				// Really Nice Point Smoothing
	glEnable(GL_TEXTURE_2D);							// Enable Texture Mapping
	glBindTexture(GL_TEXTURE_2D,texture[0]);			// Select Our Texture

	for (loop=0;loop<MAX_PARTICLES;loop++)				// Initials All The Textures
	{
		particle[loop].active=true;						// Make All The Particles Active
		particle[loop].life=1.0f;						// Give All The Particles Full Life
		particle[loop].fade=float(rand()%100)/1000.0f+0.003f;	// Random Fade Speed
		particle[loop].r=colors[(loop+1)/(MAX_PARTICLES/12)][0];	// Select Red Rainbow Color
		particle[loop].g=colors[(loop+1)/(MAX_PARTICLES/12)][1];	// Select Green Rainbow Color
		particle[loop].b=colors[(loop+1)/(MAX_PARTICLES/12)][2];	// Select Blue Rainbow Color
		particle[loop].xi=float((rand()%50)-26.0f)*10.0f;	// Random Speed On X Axis
		particle[loop].yi=float((rand()%50)-25.0f)*10.0f;	// Random Speed On Y Axis
		particle[loop].zi=float((rand()%50)-25.0f)*10.0f;	// Random Speed On Z Axis
		particle[loop].xg=0.0f;							// Set Horizontal Pull To Zero
		particle[loop].yg=-0.8f;						// Set Vertical Pull Downward
		particle[loop].zg=0.0f;							// Set Pull On Z Axis To Zero
	}

        for (int yu=0;yu<256;yu++) 
             keys[yu]=0;

	return 1;										// Initialization Went OK
}


void display()										// Here's Where We Do All The Drawing
{
	glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);		// Clear Screen And Depth Buffer

	for (loop=0;loop<MAX_PARTICLES;loop++)					// Loop Through All The Particles
	{
		if (particle[loop].active)							// If The Particle Is Active
		{
			glLoadIdentity();								// Reset The ModelView Matrix
			// Position Each Particle On The Screen, Use Zoom To Move Into The Screen

                         //Spin me crazy mode*********************** 
                         if (keys[62])
                         {
                             glRotatef(spin, 0,0,1);
                             glTranslatef(spin1, spin2,0);
                         }
                         //*****************************************

                        glTranslatef(particle[loop].x,particle[loop].y,particle[loop].z+zoom);
                         
			// Draw The Particle Using Our RGB Values, Fade The Particle Based On It's Life
			glColor4f(particle[loop].r,particle[loop].g,particle[loop].b,particle[loop].life);
			glCallList(image);									// Draw Our Particle
			particle[loop].x+=particle[loop].xi/(slowdown*1000);// Move On The X Axis By X Speed
			particle[loop].y+=particle[loop].yi/(slowdown*1000);// Move On The Y Axis By Y Speed
			particle[loop].z+=particle[loop].zi/(slowdown*1000);// Move On The Z Axis By Z Speed

			particle[loop].xi+=particle[loop].xg;			// Take Pull On X Axis Into Account
			particle[loop].yi+=particle[loop].yg;			// Take Pull On Y Axis Into Account
			particle[loop].zi+=particle[loop].zg;			// Take Pull On Z Axis Into Account
			particle[loop].life-=particle[loop].fade;		// Reduce Particles Life By 'Fade'

			if (particle[loop].life<0.0f)					// If Particle Is Burned Out
			{
				particle[loop].life=1.0f;					// Give It New Life
				particle[loop].fade=float(rand()%100)/1000.0f+0.003f;	// Random Fade Value
				particle[loop].x=0.0f;						// Center On X Axis
				particle[loop].y=0.0f;						// Center On Y Axis
				particle[loop].z=0.0f;						// Center On Z Axis
				
                                if (keys[61]) //constant explosion mode *****************
                                    {
                                        particle[loop].xi=float((rand()%50)-26.0f)*10.0f;	// Random Speed On X Axis
         				particle[loop].yi=float((rand()%50)-25.0f)*10.0f;	// Random Speed On Y Axis
	        			particle[loop].zi=float((rand()%50)-25.0f)*10.0f;	// Random Speed On Z Axis
                                    }
                                else
                                    {
                                         particle[loop].xi=xspeed+float((rand()%60)-32.0f);	// X Axis Speed And Direction
				         particle[loop].yi=yspeed+float((rand()%60)-30.0f);	// Y Axis Speed And Direction
				         particle[loop].zi=float((rand()%60)-30.0f);	// Z Axis Speed And Direction
                                    }
				particle[loop].r=colors[col][0];			// Select Red From Color Table
				particle[loop].g=colors[col][1];			// Select Green From Color Table
				particle[loop].b=colors[col][2];			// Select Blue From Color Table
			}

			// If Number Pad 8 And Y Gravity Is Less Than 1.5 Increase Pull Upwards
			if (keys[8] && (particle[loop].yg<1.5f)) { particle[loop].yg+=0.2f;
                                                                 }

			// If Number Pad 2 And Y Gravity Is Greater Than -1.5 Increase Pull Downwards
			if (keys[2] && (particle[loop].yg>-1.5f)) { particle[loop].yg-=0.2f;
                                                                   }

			// If Number Pad 6 And X Gravity Is Less Than 1.5 Increase Pull Right
			if (keys[6] && (particle[loop].xg<1.5f)) { particle[loop].xg+=0.2f;
                                                                  }

			// If Number Pad 4 And X Gravity Is Greater Than -1.5 Increase Pull Left
			if (keys[4] && (particle[loop].xg>-1.5f)) { particle[loop].xg-=0.2f;
                                                                   }

			if (keys[60])										// Tab Key Causes A Burst
			{
				particle[loop].x=0.0f;								// Center On X Axis
				particle[loop].y=0.0f;								// Center On Y Axis
				particle[loop].z=0.0f;								// Center On Z Axis
				particle[loop].xi=float((rand()%50)-26.0f)*10.0f;	// Random Speed On X Axis
				particle[loop].yi=float((rand()%50)-25.0f)*10.0f;	// Random Speed On Y Axis
				particle[loop].zi=float((rand()%50)-25.0f)*10.0f;	// Random Speed On Z Axis
                         }
		}
    }
	keys[60]=0;
        keys[8]=0;
        keys[2]=0;
        keys[6]=0;
        keys[4]=0;
	glutSwapBuffers();								// Everything Went OK
}

void
keyboard(unsigned char key, int x, int y)
{
    
    switch (key) 
    {
     case 27:
        exit(0);
	break;
     case '8': keys[8]=1; break;
     case '2': keys[2]=1; break;
     case '6': keys[6]=1; break;
     case '4': keys[4]=1; break;
     case  9 : keys[60]=1; break;       //explosion
     case 'e': keys[61]^=1; break;      //continuous explosion
     case 'c': keys[62]^=1; break;      //going crazy
     case ' ': col++;                   // Change The Particle Color
               if (col>11)	col=0;	// If Color Is To High Reset It
               break;
     case '+': if (slowdown>1.0f) slowdown-=0.1f;
               break;
     case '-': if (slowdown<4.0f) slowdown+=0.1f; 
               break;
     case 13: rainbow^=1; 
                break;
     
    }
}


void specialfunc(int key, int x, int y)
{
     switch (key) 
    {
    case GLUT_KEY_LEFT :     if (xspeed>-200) xspeed-=1.0f;
                             break;
    case GLUT_KEY_UP   :     if (yspeed<200) yspeed+=1.0f;
                             break;
    case GLUT_KEY_RIGHT:     if (xspeed<200) xspeed+=1.0f;
                             break;
    case GLUT_KEY_DOWN :     if (yspeed>-200) yspeed-=1.0f;
                             break;
    case GLUT_KEY_PAGE_UP:   zoom+=0.1f;
                             break;
    case GLUT_KEY_PAGE_DOWN: zoom-=0.1f;
                             break;
    }
}


void idleFunc()
{
  //Are we in rainbow mode*******************
  if (rainbow && (delay>25))
     { 
         delay=0;              // Reset The Rainbow Color Cycling Delay
         col++;                // Change The Particle Color
         if (col>11)	col=0;	// If Color Is To High Reset It
     }

  if (rainbow)  delay++;
  //****************************************             
  
  //spin me crazy mode**********************
  if (keys[62])    
      {
          spin+=2;
          spin1+=df1;
          spin2+=df2;
        
          if (spin>360) spin-=360;
          if ((spin1>2)||(spin1<-2)) df1=-df1;
          if ((spin2>2)||(spin2<-2)) df2=-df2;
        
          zoom+=df;
          if ((zoom<-60)||(zoom>-10)) df=-df;
      } 
   //***************************************
  
 glutPostRedisplay();
}




int main(int argc, char** argv)
{
    glutInitDisplayMode(GLUT_RGB|GLUT_DOUBLE);
    glutInitWindowPosition(50, 50);
    glutInitWindowSize(800, 600);
    glutInit(&argc, argv);
    glutCreateWindow("Particles");
    glutDisplayFunc(display);
    glutReshapeFunc(reshape);
    glutIdleFunc(idleFunc);
    glutKeyboardFunc(keyboard);
    glutSpecialFunc(specialfunc);
    InitGL();
    glutMainLoop();
    return 0;
}
